### ISSUE TITLE
Simplify app-level front end component access

### ISSUE THREAD

**Issue by marksreeves** [BorderTech/wcomponents#172]
At the moment it is a bit of a palaver to add CSS or JS customisation of individual instances of components (or groups of components) at the app level.

This allows an enterprise to tightly control appearance and function. It does, however, make it difficult to have exceptions or for individual applications which are slightly outside the regular corporate look & feel to implement their customisations. Such an app basically needs a custom theme or some rather labyrinthine Java workarounds.

I propose we make the process of front end customisation easier but in a way which allows an enterprise theme to disallow - by adding XML attribute(s).

In the first instance how about we add an XML attribute which allows a String HTML class attribute to be added to any UI component such as outlined in the following?

1. add an optional attribute (such as `@class`) to the schema for all UI components which output an element;
2. set/getHTMLClassName - maybe on AbstractWComponent - the HTML class name would be a String so if an application wanted to add multiple classNames they can simply by using a single space separated string as for HTML;
3. update renderers to output the HTML class name as an optional attribute; and
4. update default XSLT to output the string in HTML class attributes for each component.

Making this an XML attribute would allow a particularly tightly controlled theme to simply not output the XML `@class` attribute in the XSLT.

@jonathanaustin, @Joshua-Barclay, @JohnMcGuinness all comments welcome!

**Comment by marksreeves** [comment]
I am re-opening this issue as I have been lead to believe the Java API for adding application-level CSS & JavaScript is too opaque.

To add a CSS file one has to use `Headers.addHeadline` or `Headers.addUniqueHeadline`. This has a few issues:

1. it is poorly documented;
2. it is not intrinsically sensible - what is a 'headline' and why would one want one?;
3. it is not exposed on any WComponent (WApplication _may_ be suitable);
4. it seems to conflate several concepts (or the JavaDoc is misleading - see 1) in that it uses the language of headers which imply HTTP headers but applies lines to the HTML head element but the JavaDoc states it can also apply attributes directly to the HTML element but this is not exposed in ui:root and therefore is not part of the XSLT.

Taking our universal 80% rule any user of WComponents should be able to add (preferably to a WApplication) a CSS link or a JavaScript src and have the framework deal with it adequately - everything else can be passed to custom templates.

@jonathanaustin Maybe you can clarify some of these for me?

**Comment by marksreeves** [comment]
Re-open.

**Comment by JohnMcGuinness** [comment]
How about at the component level a method like `addStyleClass(String aStyleClass)`is added. And also  in order to add CSS or JS an annotation is added to the applications subclass of `WServlet`.

#249 proposes to allow configuration to be added via annotations perhaps something like the following might be intuitive for a Java developer.

```
    @StyleSheet(path="a/path/to/a/css/file.css")
    @Script(path="a/path/to/a/js/file.js")
    public class MyServlet extends WServlet {}
```

or if multiples are required

```
    @StyleSheets(@StyleSheet(path="a/path/to/a/css/file.css"), 
                 @StyleSheet(path="a/path/to/another/css/file.css"))
    @Scripts(@Script(path="a/path/to/a/js/file.js"), 
             @Script(path="a/path/to/another/js/file.js"))
    public class MyServlet extends WServlet {}
```

WDYT @marksreeves @jonathanaustin @ricksbrown

**Comment by marksreeves** [comment]
## Add XML element local names to the Java API

Whilst working on #207 I have been building a convoluted example manipulating ColumnLayout rows and cells which are deeply nested in a component which has a htmlClass applied so I have another suggestion to help with this:

**Define the schema local names as a set of String constants which are applied to components which have XML output.**

### Why? 

1. We can output the localName of each XML element as a HTML class on the transformed element  but the Java API has no way to reference these. If we add this ability then a human writing application CSS in WText (for example) could reference them directly from the Java. Even for external style sheets having a fixed API of HTML classNames would lead to much less confusion and keep us a bit more honest (see 3).
2. The component renderers could reference the same Strings when writing the XML.
3. Having API enforced HTML class names will stop me from randomly changing them for no good reason.

Maybe `AbstractWComponent` could have method `getXmlLocalName()` and possibly even `getXmlLocalNameAsHtmlClass()` which pre-pends the dot (as they are easy to forget). This would then require the local names be able to be referenced from the component class name. A helper method `getCssClass()` (or similar) could get the localName and the htmlClass (if applied) and concatenate them with the requisite dots (including replacing WHITE-SPACE with dot if trimmed  htmlClass contains any spaces).

### Example

* Take a WPanel with a ColumnLayout;
* WPanel XML has an element name ui:panel therefore the transformed output element of WPanel has a class which will always include "panel";
* ColumnLayout has an element name of ui:columnLayout and therefore the transformed wrapper will have a class "columnLayout";
* we should be able to reference these in Java;
* if we have helper methods we do not need to remember these HTML classNames.

This does not give the Java UI developer all of the convolutions involved in the transform but will at least give them somewhere to start. 

* Create a WPanel (myPanel) and optionally use `setHtmlClass("custom_class")`;
* set the myPanel's layout to ColumnLayout;
* then start with the CSS 
  `"." + myPanel.getXmlLocalName() + ["." + myPanel.getHtmlClass() + ]" > ." + XML_LOCAL_NAMES[ColumnLayout]` or however we are going to address them;
* this should output `.panel.custom_class > .columnLayout`.

``` java
WPanel myPanel = new WPanel();
myPanel.setHtmlClass("custom_class");
myPanel.setLayout (new ColumnLayout(args));

// Assume the CSS is not in an external style sheet (it should be) ...
String cssSelector =  + "." + myPanel.getXmlLocalName() + 
    "." + myPanel.getHtmlClass() + 
    " > ." + myPanel.getLayout().getXmlLocalName();
// you can see why getXmlLocalNameAsHtmlClass would be handy 
// and getCssClass() would be even handier!

// Optionally add more to the selector to look at cells or cell content.
// eg  including " > .row > .cell "; would get right down to the columns. 

String css = cssSelector + "{/* your css here */}";

WText cssText = new WText("<style type='text/css'>" + css + "</style>");
cssText.setEncodeText(false);
add(cssText); //add it somewhere, anywhere really...
```

**Comment by jonathanaustin** [comment]
Adding custom `js` and `css` is very convoluted in `WComponents`.

To greatly simplify this `WApplication` can be changed to cater for custom `js` and `css` resources.

The new api would look like:-

```java

public class MyApp extends WApplication {
....
    MyApp(){
        ....
        // Load as a static resource from the classpath. 
        addJsFile("/js/my.js");
        addCssFile("/css/my.css");
        ....
        // Load from a URL
        addJsUrl("http://example.com/my.js");
        addCssUrl("http://example.com/my.css");
        ....
    }
....
}
```
`WApplication` then handles loading these resources into the browser to be used by the application.

Being able to easily load custom `css` works well with the ability to set a custom `style` class on any component using `setHtmlClass(String)`.

**Comment by marksreeves** [comment]
@jonathanaustin I believe PR #333 was the last item we needed for this so you could close the issue. If you are waiting on your WTemplate delivery to close this issue don't forget #94 which will be completed by that work.

**Comment by marksreeves** [comment]
Apropos my previous idea of having a map of WComponent to XML element to HTML class attribute value based on the local name of the XML element. This was quite rightly quashed. We still need a "public API" for CSS but it should be by convention (as per other UI toolkits - dojo etc).

I have put in place an XSLT helper template attach local-name to each component. This has other weaknesses. So my current question is: should we move away from this and have a published convention of class attribute values? 

Consider all of the different ways we have to make rows and columns - WRow/WColumn (local names row and column); ColumnLayout (local names columnLayout and cell); GridLayout (local names gridLayout and cell with an optional intermediary to create rows which I only recently added); WCheckBoxSelect and WRadioButtonSelect where the column layout is determined by an attribute. What we want is a single class value to indicate a row and a single value to indicate a column.
